File: src/main/scala/com/gitlab/dhorman/cryptotrader/util/RequestLimiter.scala
package com.gitlab.dhorman.cryptotrader.util

import scala.concurrent.duration._

class RequestLimiter(allowedRequests: Int, perInterval: Duration) {
  private var reqCountFromFirst: Long = 0
  private var firstReqExecTime: Long = 0
  private var lastReqExecTime: Long = 0

  def get(): Duration = this.synchronized {
    val currReqTime = System.currentTimeMillis

    if (currReqTime - lastReqExecTime > perInterval.toMillis) {
      reqCountFromFirst = 0
      firstReqExecTime = currReqTime
      lastReqExecTime = currReqTime
    } else {
      val offset = perInterval.mul(reqCountFromFirst / allowedRequests).toMillis
      lastReqExecTime = firstReqExecTime + offset
    }

    reqCountFromFirst += 1

    val delay = {
      val d = lastReqExecTime - currReqTime
      if (d <= 0) 0 else d
    }

    Duration(delay, MILLISECONDS)
  }
}

File: src/main/kotlin/com/gitlab/dhorman/cryptotrader/util/RequestLimiter.kt
package com.gitlab.dhorman.cryptotrader.util

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.Duration

class RequestLimiter(private val allowedRequests: Int, private val perInterval: Duration) {
    private val lock = Mutex()
    private var reqCountFromFirst = 0L
    private var firstReqExecTime = 0L
    private var lastReqExecTime = 0L

    suspend fun get(): Duration = lock.withLock {
        val currReqTime = System.currentTimeMillis()

        if (currReqTime - lastReqExecTime > perInterval.toMillis()) {
            reqCountFromFirst = 0
            firstReqExecTime = currReqTime
            lastReqExecTime = currReqTime
        } else {
            val offset = perInterval.multipliedBy(reqCountFromFirst / allowedRequests).toMillis()
            lastReqExecTime = firstReqExecTime + offset
        }

        reqCountFromFirst += 1

        val delay = run {
            val d = lastReqExecTime - currReqTime
            if (d <= 0) 0 else d
        }

        Duration.ofMillis(delay)
    }
}

